파이썬 머신러닝 완벽 가이드 개정2판
ch8. 7
 p.538-549

# 8. 텍스트 분석
## 07. 문서 군집화 소개와 실습 (Opinion Review 데이터 세트)

### [문서 군집화 개념]
문서 군집화 : 비슷한 텍스트 구성의 문서를 군집화하는 것
- 학습 데이터 세트가 필요 없는 비지도 학습 기반으로 동작

### [Opinion Review 데이터 세트를 이용한 문서 군집화 수행하기]
UCI 머신러닝 리포지토리에 있는 Opinion Review 데이터 세트
- 51개의 텍스트 파일로 구성
- 호텔, 자동차, 전자제품 사이트에서 가져온 리뷰 문서

In [29]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [30]:
import pandas as pd
import glob, os
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.cluster import KMeans

# 필수 NLTK 데이터 다운로드
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [31]:
import pandas as pd
import glob, os
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.cluster import KMeans

# 디렉터리
path = r'/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics'
# path로 지정한 디렉터리 밑에 있는 모든 .data 파일들의 파일명을 리스트로 취합
all_files = glob.glob(os.path.join(path, '*.data'))
filename_list = []
opinion_text = []

# 개별 파일들의 파일명은 filename_list 리스트로 취합,
# 개별 파일들의 파일 내용은 DataFrame 로딩 후 다시 string으로 변환하여 opinion_text 리스트로 취합
for file_ in all_files:
    # 개별 파일을 읽어서 DataFrame으로 생성
    df = pd.read_table(file_, index_col=None, header=0, encoding='latin1')
    # 절대 경로로 주어진 파일명을 가공. Linux에서 수행 시에는 아래 \\를 / 변경.
    # 맨 마지막 .data 확장자도 제거
    filename_ = file_.split('\\')[-1]
    filename = filename_.split('.')[0]

    # 파일명 리스트와 파일 내용 리스트에 파일명과 파일 내용을 추가
    filename_list.append(filename)
    opinion_text.append(df.to_string())

# 파일명 리스트와 파일 내용 리스트를 DataFrame으로 생성
document_df = pd.DataFrame({'filename': filename_list, 'opinion_text': opinion_text})
document_df.head()

filename  \
0   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/accuracy_garmin_nuvi_255W_gps   
1     /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_netbook_1005ha   
2       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/comfort_honda_accord_2008   
3  /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/bathroom_bestwestern_hotel_sfo   
4      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  
0                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...  
1                                                                                                                                                                                                                                                                                                                  6GHz 533FSB cpu, glossy display, 3, Cell 23Wh Li, ion Battery  , and a 1 .\n0                                                                                                                                                                                                                                                                                              Not to mention that as of now...  
2                                                                                                                                                                                                                                                                                                                                                                                                Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...  
3                                                                                                                                                                                                                                                The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and 

-> 각 파일 이름 (filename) 자체만으로 의견(opinion)의 텍스트(text)가 어떠한 제품/서비스에 대한 리뷰인지 알 수 있음

문서를 TF-IDF 형태로 피처 벡터화하기
- LemNormalize() 함수 : 어근 변환 수행

In [32]:
# 텍스트 정규화 및 토큰화
def LemNormalize(text):
    wordnet_lemmatizer = WordNetLemmatizer()
    words = word_tokenize(text)  # nltk 토큰화 적용
    return [wordnet_lemmatizer.lemmatize(word) for word in words]


# 빈 문서 제거
document_df = document_df.dropna(subset=['opinion_text'])

# Ensure the 'opinion_text' column contains strings
document_df['opinion_text'] = document_df['opinion_text'].astype(str)

document_df = document_df[document_df['opinion_text'].str.len() > 0]

# TF-IDF 벡터화 (stop_words 제거, min_df 완화)
tfidf_vect = TfidfVectorizer(tokenizer=None, preprocessor=None,
                             ngram_range=(1,2), min_df=0.01, max_df=0.95)
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

문서별 텍스트가 TF-IDF 변환된 피처 벡터화 행렬 데이터에 대해서 군집화를 수행 -> 어떤 문서끼리 군집되는지 확인하기
- 군집화 기법 : K-평균 적용
- 최대 반복 횟수 max_iter = 10000

In [33]:
# KMeans 클러스터링
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0, n_init='auto')
km_cluster.fit(feature_vect)

cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

document_df 에 'cluster_label' 칼럼을 추가해 저장하기

In [34]:
document_df['cluster_label'] = cluster_label
document_df.head()

filename  \
0   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/accuracy_garmin_nuvi_255W_gps   
1     /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_netbook_1005ha   
2       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/comfort_honda_accord_2008   
3  /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/bathroom_bestwestern_hotel_sfo   
4      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  \
0                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
1                                                                                                                                                                                                                                                                                                                  6GHz 533FSB cpu, glossy display, 3, Cell 23Wh Li, ion Battery  , and a 1 .\n0                                                                                                                                                                                                                                                                                              Not to mention that as of now...   
2                                                                                                                                                                                                                                                                                                                                                                                                Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
3                                                                                                                                                                                                                                                The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter 

군집화 결과 확인해보기

- cluster_label=0 인 데이터 세트

In [35]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
2        /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/comfort_honda_accord_2008   
8        /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/comfort_toyota_camry_2007   
19   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/gas_mileage_toyota_camry_2007   
21      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/interior_honda_accord_2008   
15      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/interior_toyota_camry_2007   
16       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/mileage_honda_accord_2008   
27   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/performance_honda_accord_2008   
28       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/quality_toyota_camry_2007   
37         /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/seats_honda_accord_2008   
49  /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/transmission_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
2                                                                                                                                                                                                                                                                                                                                                                                                 Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
8                                                                                                                                                   Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                        Seats are fine, in fact of all the smaller sedans this is the most comfortable I found for the price as I am 6', 2  and 250# .\n1                                                                                                                                                                                     Great gas mileage and comfortable on long trips ...   
19                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                                                                                                                                                  ...   
21                                                                                                                                                             

-> cluster #0은 호텔에 대한 리뷰로 군집화됨

In [36]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
0   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/accuracy_garmin_nuvi_255W_gps   
6      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_amazon_kindle   
4      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_ipod_nano_8gb   
1     /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_netbook_1005ha   
11   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/display_garmin_nuvi_255W_gps   
13  /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/eyesight-issues_amazon_kindle   
12              /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/features_windows7   
20        /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/keyboard_netbook_1005ha   
18       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/navigation_amazon_kindle   
31     /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/performance_netbook_1005ha   
38    /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/screen_garmin_nuvi_255W_gps   
34           /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/screen_ipod_nano_8gb   
39          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/screen_netbook_1005ha   
44       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/size_asus_netbook_1005ha   
45            /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/sound_ipod_nano_8gb   
42     /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/speed_garmin_nuvi_255W_gps   
43                 /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/speed_windows7   
50   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/updates_garmin_nuvi_255W_gps   
48            /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/video_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
0                                                                                                                                                                                                                                                                  , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
6                                                                                                                                                                                                                After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                             

-> Cluster #1 : 킨들, 아이팟, 넷북 등의 포터블 전자기기 및 주요 구성요소에 대한 리뷰로 군집화됨

In [37]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
3    /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/bathroom_bestwestern_hotel_sfo   
5             /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/buttons_amazon_kindle   
7   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/directions_garmin_nuvi_255W_gps   
9               /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/fonts_amazon_kindle   
10          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/food_holiday_inn_london   
23           /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/food_swissotel_chicago   
22       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/free_bestwestern_hotel_sfo   
17   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/location_bestwestern_hotel_sfo   
14      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/location_holiday_inn_london   
29    /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/parking_bestwestern_hotel_sfo   
26              /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/price_amazon_kindle   
24         /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/price_holiday_inn_london   
25          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/room_holiday_inn_london   
30      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/rooms_bestwestern_hotel_sfo   
35          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/rooms_swissotel_chicago   
32   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/satellite_garmin_nuvi_255W_gps   
36    /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/service_bestwestern_hotel_sfo   
33       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/service_holiday_inn_london   
40  /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/service_swissotel_hotel_chicago   
41      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/staff_bestwestern_hotel_sfo   
46          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/staff_swissotel_chicago   
47       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/voice_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
3                                                                                                                                                                                                                                                 The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...   
5                                                                                                                                              I thought it would be fitting to christen my Kindle with the Stephen King novella UR, so went to the Amazon site on my computer and clicked on the button to buy it .\n0                                        

-> Cluster #2 : 주로 차량용 내비게이션 리뷰로 군집화됨

In [38]:
document_df[document_df['cluster_label']==3].sort_values(by='filename')

Empty DataFrame
Columns: [filename, opinion_text, cluster_label]
Index: []

-> Cluster #3 : 킨들 리뷰가 한 개 섞여 있지만, 대부분 호텔에 대한 리뷰로 군집화됨

In [39]:
document_df[document_df['cluster_label']==4].sort_values(by='filename')

Empty DataFrame
Columns: [filename, opinion_text, cluster_label]
Index: []

-> Cluster #4 : 토요타, 혼다 등의 자동차에 대한 리뷰로 군집화됨

=> 군집 개수가 약간 많게 설정돼 있어서 세분화되어 군집화된 경향이 있음

중심 개수를 5개에서 3개로 낮춰서 3개 그룹으로 군집화한 뒤 결과 확인해보기

In [40]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_

# 소속 클러스터를 cluster_label 칼럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by='cluster_label')

filename  \
2         /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/comfort_honda_accord_2008   
8         /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/comfort_toyota_camry_2007   
15       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/interior_toyota_camry_2007   
28        /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/quality_toyota_camry_2007   
27    /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/performance_honda_accord_2008   
21       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/interior_honda_accord_2008   
19    /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/gas_mileage_toyota_camry_2007   
16        /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/mileage_honda_accord_2008   
49   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/transmission_toyota_camry_2007   
37          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/seats_honda_accord_2008   
13    /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/eyesight-issues_amazon_kindle   
11     /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/display_garmin_nuvi_255W_gps   
6        /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_amazon_kindle   
12                /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/features_windows7   
31       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/performance_netbook_1005ha   
0     /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/accuracy_garmin_nuvi_255W_gps   
20          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/keyboard_netbook_1005ha   
18         /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/navigation_amazon_kindle   
45              /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/sound_ipod_nano_8gb   
44         /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/size_asus_netbook_1005ha   
34             /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/screen_ipod_nano_8gb   
43                   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/speed_windows7   
50     /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/updates_garmin_nuvi_255W_gps   
48              /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/video_ipod_nano_8gb   
39            /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/screen_netbook_1005ha   
42       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/speed_garmin_nuvi_255W_gps   
1       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_netbook_1005ha   
4        /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/battery-life_ipod_nano_8gb   
38      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/screen_garmin_nuvi_255W_gps   
5             /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/buttons_amazon_kindle   
14      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/location_holiday_inn_london   
10          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/food_holiday_inn_london   
30      /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/rooms_bestwestern_hotel_sfo   
29    /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/parking_bestwestern_hotel_sfo   
25          /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/room_holiday_inn_london   
26              /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/price_amazon_kindle   
24         /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/price_holiday_inn_london   
17   /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/location_bestwestern_hotel_sfo   
22       /content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/free_bestwestern_hotel_sfo   
23           /content/drive/MyDrive/2025-1/2025-1

### [군집별 핵심 단어 추출하기]
각 군집을 구성하는 핵심 단어가 어떤 것이 있는지 확인하기

KMeans 객체의 clusters_centers_
- 각 군집을 구성하는 단어 피처가 군집의 중심을 기준으로 얼마나 가깝게 위치해 있ㄴ느지
- 배열 값으로 제공, 행은 개별 군집, 열은 개별 피처
- 각 배열 내의 값 : 개별 군집 내의 상대 위치를 숫자 값으로 표현한 일종의 좌표 값

In [41]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape : ', cluster_centers.shape)
print(cluster_centers)

cluster_centers shape :  (3, 66040)
[[0.00071021 0.         0.         ... 0.         0.         0.        ]
 [0.00065802 0.00109736 0.         ... 0.00238505 0.00132443 0.00117224]
 [0.00299631 0.         0.00087601 ... 0.         0.         0.        ]]


- (3, 66040) 배열 : 군집이 3개, word 피처가 66040개로 구성됨
- 0에서 1까지의 값 : 1에 가까울수록 중심과 가까운 값


cluster_centers_ 속성값을 이용해 각 군집별 핵심 단어 찾아보기
- cluster_centers_ 속성 : 넘파이의 ndarray
- 위치 인덱스 값 : 핵심 단어 피처의 이름을 출력하기 위해 필요

get_cluster_details() 함수 생성
- cluster_centers_ 배열 내에서 가장 값이 큰 데이터의 위치 인덱스를 추출
- 해당 인덱스를 이용해 핵심 단어 이름과 그때의 상대 위치 값을 추출
- cluster_details라는 Dict 객체 변수에 기록하고 반환하기

In [49]:
# 군집별 top n 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명을 반환함
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
    cluster_details = {}

    # cluster_centers array의 값이 큰 순으로 정렬된 인덱스 값을 반환
    # 군집 중심점별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:, ::-1]

    # 개별 군집별로 반복하면서 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명 입력
    for  cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster'] = cluster_num

        # cluster_centers_.argsort()[:, ::-1]로 구한 인덱스를 이용해 top_n 피처 단어를 구함
        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [ feature_names[ind] for ind in top_feature_indexes ]

        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()

        # cluster_details 딕셔너리 객체에 개별 군집별 핵심단어와 중심위치 상댓값, 해당 파일명 입력
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()

        cluster_details[cluster_num]['filenames'] = filenames

    return cluster_details

get_cluster_details() 호출 : dictionary를 원소로 가지는 리스트인 cluster_details를 반환
- cluster_details : 개별 군집번호, 핵심 단어, 핵심 단어 중심 위치 상댓값, 파일명 속성 값 정보
- 이를 보기 좋게 print_cluster_details() 함수 제작

In [43]:
def print_cluster_details(cluster_details):
    for  cluster_num, cluster_detail in cluster_details.items():
        print('##### Cluster {0}'.format(cluster_num))
        print('Top features:', cluster_detail['top_features'])
        print('Reviews 파일명 : ', cluster_detail['filenames'][:7])
        print(' ============================================== ')

두 함수 호출하기

In [50]:
feature_names = tfidf_vect.get_feature_names_out()

cluster_details = get_cluster_details(cluster_model=km_cluster, cluster_data=document_df, feature_names=feature_names, clusters_num=3, top_n_features=10)

print_cluster_details(cluster_details)


##### Cluster 0
Top features: ['interior', 'mileage', 'seats', 'gas', 'comfortable', 'gas mileage', 'the interior', 'transmission', 'car', 'performance']
Reviews 파일명 :  ['/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/comfort_honda_accord_2008', '/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/comfort_toyota_camry_2007', '/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/interior_toyota_camry_2007', '/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/mileage_honda_accord_2008', '/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/gas_mileage_toyota_camry_2007', '/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/interior_honda_accord_2008', '/content/drive/MyDrive/2025-1/2025-1 ESAA OB/과제/data/topics/performance_honda_accord_2008']
##### Cluster 1
Top features: ['battery', 'screen', 'battery life', 'the screen', 'life', 'keyboard', 'the battery', 'video', 'speed', 'the keyboard']
Reviews 파일명 :  ['/content/drive/